src/cpp/session/modules/SessionDependencies.cpp

```diff
@@ -310,106 +311,97 @@ Error installDependencies(const json::JsonRpcRequest& request,
+      // treatment, since comment headers are used to show the currently executing section of a
+      // script.
+      script += "# " + dep.name + " -------------\n\n";
+
+      // If there's more than one dependency, indicate which one is being installed.
+      if (deps.size() > 1)
+      {
+         script += "cat('\\n[" + 
+            safe_convert::numberToString(i + 1) + 
+            "/" + 
+            safe_convert::numberToString(deps.size()) +
+            "] Installing " + dep.name + "...\\n\\n')\n";
+      }
+
       if (dep.location == kCRANPackageDependency)
       {
-         if (dep.source)
-            cranSourcePackages.push_back("'" + dep.name + "'");
-         else
-            cranPackages.push_back("'" + dep.name + "'");
+         // Build install command for CRAN. We specify lock = TRUE (here and elsewhere) to ensure
+         // that the package directory is locked during installation; since this will run in the
+         // background we want reduce the odds of corruption via a competing package install attempt
+         // in another process.
+         script += "utils::install.packages('" + dep.name + "', " +
+                "repos = '"+ module_context::CRANReposURL() + "'";
+
+         if (dep.source) 
+         {
+            // Install from source if requested
+            script += ", type = 'source'";
+         }
+
+         script += ", lock = TRUE)";
       }
       else if (dep.location == kEmbeddedPackageDependency)
       {
          EmbeddedPackage pkg = embeddedPackageInfo(dep.name);
-         if (!pkg.empty())
-            embeddedPackages.push_back(pkg.archivePath);
+
+         // Build install command for bundled archive
+         script += "utils::install.packages('" + pkg.archivePath + 
+            "', repos = NULL, type = 'source', lock = TRUE)";
       }
-   }
 
-   // build install command
-   std::string cmd("{ " + module_context::CRANDownloadOptions() + "; ");
-   if (!cranPackages.empty())
-   {
-      std::string pkgList = boost::algorithm::join(cranPackages, ",");
-      cmd += "utils::install.packages(c(" + pkgList + "), " +
-             "repos = '"+ module_context::CRANReposURL() + "'";
-      cmd += ");";
+      script += "\n\n";
    }
-   if (!cranSourcePackages.empty())
-   {
-      std::string pkgList = boost::algorithm::join(cranSourcePackages, ",");
-      cmd += "utils::install.packages(c(" + pkgList + "), " +
-             "repos = '"+ module_context::CRANReposURL() + "', ";
-      cmd += "type = 'source');";
-   }
-   for (const std::string& pkg : embeddedPackages)
-   {
-      cmd += "utils::install.packages('" + pkg + "', "
-                                      "repos = NULL, type = 'source');";
-   }
-   cmd += "}";
 
-   // build args
-   std::vector<std::string> args;
-   args.push_back("--slave");
+   jobs::ScriptLaunchSpec installJob(
+         // Job name: just the package name if we have one 
```

Comment:
Could we also include associated context for package installation (if available)? E.g. something like:

```
R Notebooks: Install rmarkdown
```

Reply:
Good idea! Done.